WordPress/gutenberg
# Editor: Implement meta as custom source

Closes #16282
Related (alternative to): #16075

This pull request seeks to explore an implementation of custom sources.

**Implementation Notes:**

The implementation inherits a fair bit from #16075 (notably "last changes" tracking), but it tries to leverage a flow where sourced attributes are updated (as applicable) and applied prior to any blocks reset from the editor module. Furthermore, per discussion at https://github.com/WordPress/gutenberg/pull/16075#discussion_r298365062, they are implemented as standard store controls, where the benefit of this approach is in increased flexibility of the implementation by avoiding the need to determine a specific set of arguments which would apply for all potential custom source implementations. Instead, each implementation can simply yield controls (such as `select` from `@wordpress/data-controls`) to retrieve whichever data they depend upon.

It diverges from some alternative proposals (https://github.com/WordPress/gutenberg/issues/16282#issuecomment-507423808) due to:

- Not having well-defined "start" and "end" cycles from within the context of a block editor
- The difficulty and overhead in keeping redundant data in sync for "sourced" data, where this implementation can derive directly from post meta (rather than to carve a separate state which needs to be kept in sync with the post)
- In the alternative, not being able to eliminate selector-time source derivation (as of the changes here, `getBlockAttributes` returns a static reference)

As to the proposed "Block Sources API", I'm not very attached to the specific set of arguments currently passed to `apply`, `applyAll`, and `update`. The idea for `applyAll` came about as a result of [a performance concern](https://github.com/WordPress/gutenberg/pull/16075#discussion_r298489530) for repeated data access, but it is not strictly necessary (`apply` and `update` alone would be nicely complementing, though similarly we could explore to add an `updateAll`). I didn't seek to make this publicly-extensible for the moment, though it is designed to allow for it in the future after some internal validation of the API.

**Remaining Tasks:**

- [ ] Unit tests to be written, if implementation deemed acceptable.

**Testing Instructions:**

Verify that updating a block with an attribute sourced by a meta attribute reflects the update (it is restored after a save, and it applies to all other blocks which source from the same meta).

As an example, I think this demo plugin should still work: https://gist.github.com/pento/19b35d621709042fc899e394a9387a54

Caveats:

- Block validation fails if the meta value is persisted into post content (#4989). I aim to tackle this separately, where validation occurs as a separated process from the default blocks `parse` behavior and runs only after the meta values are applied.
- There seems to be an issue where other blocks sourced from the same meta do not re-render immediately. Clicking to select those other blocks will trigger the render and reflect the update. I suspect there may be an issue with either the editor's async rendering, or how we choose to memoize these blocks' components rendering.
